### PR TITLE
Adds schema caching capabilities (5s by default)

### PR DESCRIPTION
**Pull request by flovilmart**
Adds caching of the schema.

- configure with schemaCacheTTL, or PARSE_SERVER_SCHEMA_CACHE_TTL options.
- 5s by default.
- put 0s or false to disable

A new cache is created for each request, scoped with a prefixed key, so no information is leaked across requests.
This could be further optimized later on.

### CODE REVIEW

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

File: src/Controllers/SchemaCache.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+const CACHED_KEYS = "__CACHED_KEYS";
```
Comment (steven-supersolid) [inline src/Controllers/SchemaCache.js]: Unused?

Comment (flovilmart) [inline src/Controllers/SchemaCache.js]: right :)

File: src/Controllers/SchemaCache.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  setAllClasses(schema) {
+    if (this.ttl <= 0) {
+      return Promise.resolve(null);
+    }
+    this.cache.put(SCHEMA_CACHE_PREFIX+MAIN_SCHEMA, schema, this.ttl);
+  }
+
+  setOneSchema(className, schema) {
```
Comment (steven-supersolid) [inline src/Controllers/SchemaCache.js]: If we set one schema/class it won't be available in getAllClasses() - is this a concern?

Comment (flovilmart) [inline src/Controllers/SchemaCache.js]: Not really, at that moment, the cache is just on the top of the DB queries to make things simple.

File: src/ParseServer.js [hunk 3]
```diff
@@ -132,6 +133,7 @@ class ParseServer {
     expireInactiveSessions = true,
     verbose = false,
     revokeSessionOnPasswordReset = true,
+    schemaCacheTTL = -1, // -1 = no cache
```
Comment (steven-supersolid) [inline src/ParseServer.js]: We now have NaN and -1 as special TTL values, should we put these as constants somewhere?

Comment (flovilmart) [inline src/ParseServer.js]: I didn't know we had NaN TTL's where is that?

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
Closes #1869

**Comment by drew-gross** [comment]
Boolean parameters kinda suck, it's so hard to tell at the call site what they are for. Can you change these functions to use object-of-parameters style? `reloadDate({ force })` etc.

@blacha can you look at how he is instantiating the cache adapter? I think it was designed be only be instantiated once.

**Comment by flovilmart** [comment]
I'll update the style. Initially I didn't go for the CacheController because I needed / wanted a per request cache, and not a global one in order to minimize side effects and minimize the number of changes on the Controllers to take that into account.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by benitech** [comment]
This sets cache for "_Role, _User and _SCHEMA".
How can i add caching for another class that i created? Possible to add the new class name somewhere in index.js?

**Pull request merged**